### PR TITLE
Implement merging rules

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
## Overview
Implements most of the merging rules from the spreadsheet

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3638

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by lukasz-golebiewski** [comment]
bors try

**Comment by lukasz-golebiewski** [comment]
bors try

**Comment by lukasz-golebiewski** [comment]
bors try

**Comment by lukasz-golebiewski** [comment]
bors try

File: casper/src/test/scala/coop/rchain/casper/EstimatorHelperTest.scala [hunk 1]
```diff
@@ -89,32 +89,38 @@ class EstimatorHelperTest
...
       }
   }
 
-  it should "conflict if their deploys contain same channel in paymentLog" in withStorage {
+  it should "conflict if their deploys conflict in the paymentLog" in withStorage {
     implicit blockStore => implicit blockDagStorage =>
       testConflict[Task] { deploy =>
         deploy.copy(paymentLog = Seq(produce(ByteString.copyFromUtf8(channelsHash))))
+      } { deploy =>
+        deploy.copy(paymentLog = Seq(consume(ByteString.copyFromUtf8(channelsHash))))
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/EstimatorHelperTest.scala:116]: Please add a case doing a cross test (payment conflicts with deploy)

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala [hunk 2]
```diff
@@ -236,43 +233,48 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
...
+  it should "respect mergeability rules when merging blocks" in effectTest {
+    mergeabilityCases.toList
+      .map {
+        case ((name, desc), y) =>
+          val key = s"$name $desc"
+          y.attempt.map(_.fold(_ => s"$key FAILED", _ => key))
+      }
+      .parSequence
+      .map(_.filter(_.contains("FAILED")))
+      .map { r =>
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:256]: subsequent maps can be squashed
map after sequence is traverse

Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:256]: I guess if we're just interested in failure cases (which is a greate choice!) we could just filter for Left-s

Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:256]: Notice that the `desc` string is hellla long and will obscure output. I see two ways out of that:

1) make the `name`-s unique, add a validation in the test-validating test that forces that
2) add whitespace, so that the list of failures looks like:

```
$name FAILED:

$desc

$name FAILED:

$desc

...

```

Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:256]: I also can't wait for the test case self-validation we've promised each other (encoding the conditions stated in `desc` strings as simple rules validating that a given example meets the preconditions), e.g.:

```
//should fail a precondition because "4" and "!" don't have same polarity
samePolarityMerge("4X !X", S0, F_, Nil)`

//should fail a precondition because S0 is not a receive as dictated by 4X
samePolarityMerge("4X 4!", S0, F_, Nil)`
```

Are you going to add it here, or in the next PR? :)

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala [hunk 3]
```diff
@@ -236,43 +233,48 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
     "4! (4!)"   -> VOLATILE_EVENT             -> merges(F_, F1 | S1, S0),
     "4! (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(F_, C_ | S0, S0),
     "4X (4!)"   -> VOLATILE_EVENT             -> merges(F_, F_ | S0, Nil),
-    "4X (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflictsForNow(F_, C_ | S0, Nil),
-    "C! (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflictsForNow(C_, C_ | S0, S0),
-    "CX (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflictsForNow(C_, C_ | S0, Nil),
+    "4X (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(F_, C_ | S0, Nil),
+    "C! (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(C_, C_ | S0, S0),
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: interesting... This should merge, because `C! C!` merges

Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: oh I see why, the test case is wrong - should be `merges(C_, C1 | S1, S0)`, or the right C_ eats the base send as well

**Comment by lukasz-golebiewski** [comment]
bors try

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 2 above)
Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:256]: given that I've just spotted a case where PERSISTENT_COULD_NOT_MATCH was marked as `conflicts` (should be merge for all PERSISTENT_COULD_NOT_MATCH) - having the self-validation would help reviewing this.

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala [hunk 4]
```diff
@@ -146,70 +146,69 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
...
-    "!4 C!"     -> HAD_ITS_MATCH              -> conflictsForNow(S0, C_, F0 | S1),
-    "!4 (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(S0, C_ | S0, F0),
-    "!4 (!C)"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(S0, S0 | C_, F0),
-    "!C !C"     -> COULD_MATCH_SAME_TRUE      -> conflictsForNow(S0, S0, C_),
-    "!C (!C)"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(S0, C_ | S1, C0),
-    "!C 4X"     -> HAD_ITS_MATCH              -> conflictsForNow(S0, F_, C_),
-    "!C 4!"     -> HAD_ITS_MATCH              -> conflictsForNow(S0, F_, C0 | S1),
+    "!4 4!!"    -> HAD_ITS_MATCH              -> merges(S0, F_, F0 | R1),
+    "!4 !!X"    -> SAME_POLARITY_MERGE        -> merges(S0, R1, F0),
+    "!4 !!4"    -> SAME_POLARITY_MERGE        -> conflicts(S0, R1, F_ | F1),
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: `SAME_POLARITY_MERGE        -> conflicts`
*jazz music stops*

The test case should be `merges(S0, R1, F0 | F1)`

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala [hunk 5]
```diff
@@ -146,70 +146,69 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
...
+    "!4 CX"     -> HAD_ITS_MATCH              -> merges(S0, C_, F_),
+    "!4 C!"     -> HAD_ITS_MATCH              -> merges(S0, C_, F0 | S1),
+    "!4 (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(S0, C_ | S1, F0),
+    "!4 (!C)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(S0, S1 | C_, F0),
+    "!C !C"     -> COULD_MATCH_SAME_TRUE      -> merges(S0, S0, C_),
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: this should be `COULD_MATCH_SAME_FALSE`. The FALSE refers to whether it conflicts or not. It used to be `same` vs `diff`. Maybe the cases should be named

COULD_MATCH_SAME_CONFLICT and COULD_MATCH_SAME_MERGE

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 6]
```diff
@@ -91,32 +127,132 @@ object EstimatorHelper {
...
+
+  trait Polarity
+  case object Send    extends Polarity
+  case object Receive extends Polarity
+
+  trait Cardinality
+  case object Linear    extends Cardinality
+  case object NonLinear extends Cardinality
+
+  object TuplespaceEvent {
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: That's a lot of new code that we should extract to a separate file, and think of ways to separate it from EstimatorHelper for better testability maybe (less coupled setup -> faster tets)

**Review by ArturGajowy — CHANGES_REQUESTED** [review]
Sharing what I have now, will add more later

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 7]
```diff
@@ -35,54 +37,88 @@ object EstimatorHelper {
...
+                s"Blocks ${PrettyPrinter
+                  .buildString(b1.blockHash)} and ${PrettyPrinter
+                  .buildString(b2.blockHash)} don't conflict."
+              )
+            }
+      } yield conflicts
     }
 
-  private def isVolatile(comm: COMM, consumes: Seq[Consume], produces: Seq[Produce]) =
+  private[this] def containConflictingEvents(
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:79]: typo?

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 8]
```diff
@@ -91,32 +127,132 @@ object EstimatorHelper {
...
+      b: BlockEvents
+  ): Map[Blake2b256Hash, Set[TuplespaceEvent]] =
+    b match {
+      case (produces, consumes, comms) =>
+        val nonPersistentProducesInComms = comms.flatMap(_.produces).filterNot(_.persistent)
+        val nonPersistentConsumesInComms = comms.map(_.consume).filterNot(_.persistent)
+        val freeProduces                 = produces.filterNot(nonPersistentProducesInComms.contains)
+        val freeConsumes                 = consumes.filterNot(nonPersistentConsumesInComms.contains)
+        val produceEvents = freeProduces
+          .map(TuplespaceEvent.from(_))
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: formatting?

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 9]
```diff
@@ -91,32 +127,132 @@ object EstimatorHelper {
...
+  implicit class TuplespaceEventOps(val ev: TuplespaceEvent) extends AnyVal {
 
-      ancestorChannels = ancestorProduceChannels ++ ancestorConsumeChannels ++ ancestorCommChannels
-      res              = ancestorChannels.map(_.toByteString)
-    } yield res
+    private[casper] def conflicts(other: TuplespaceEvent): Boolean =
+      if (ev.incoming.polarity == other.incoming.polarity)
+        (for {
+          a <- ev.matched
+          b <- other.matched
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: are you happy with the `a` `b` names?

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 10]
```diff
@@ -91,32 +127,132 @@ object EstimatorHelper {
...
+        } yield a == b && b.cardinality == Linear).getOrElse(false)
+      else
+        !(
+          ev.incoming.cardinality == Linear && other.incoming.cardinality == Linear && (ev.matched != None || other.matched != None) ||
+
+            ev.incoming.cardinality == NonLinear && other.matched != None ||
+            other.incoming.cardinality == NonLinear && ev.matched != None
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: would it be possible to new line this past every `&&` case?

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 4 above)
Comment (lukasz-golebiewski) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: I beg to differ. Here's how is should look like:
```
"!4 !!4"    -> COULD_MATCH_SAME_TRUE      -> conflicts(S0, R1, F_ | F1),
"!4 !!4"    -> COULD_MATCH_SAME_FALSE     -> merges(S0, R1, F0 | F1),
```

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 5 above)
Comment (lukasz-golebiewski) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: Good catch! Renaming sounds good

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 6 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: Good idea, I would rather do that in a follow up PR though if you don't mind

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 7 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:79]: Can't see it

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 8 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: Not sure what you mean, this code has been auto-formatted

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 9 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: I am, but I could use better names for readability's sake. What would you think about `leftMatched` and `rightMatched`?

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 10 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: `&&` precedes `||` so this would make things less readable

File: casper/src/test/scala/coop/rchain/casper/EstimatorHelperTest.scala (hunk 1 above)
Comment (lukasz-golebiewski) [reply casper/src/test/scala/coop/rchain/casper/EstimatorHelperTest.scala:116]: Will do

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 2 above)
Comment (lukasz-golebiewski) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:256]: I would opt for a new PR

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 5 above)
Comment (lukasz-golebiewski) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: Done

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 3 above)
Comment (lukasz-golebiewski) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: Done

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 7 above)
Comment (dzajkowski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:79]: `containConflictingEvents` -> `containsConflictingEvents` ?

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 8 above)
Comment (dzajkowski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]:
```
val produceEvents = freeProduces
          .map(TuplespaceEvent.from(_))
```
```
val produceEvents = freeProduces.map(TuplespaceEvent.from(_))
```

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 9 above)
Comment (dzajkowski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: thisMatched, otherMatched? there is no left/right in this context imo

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 10 above)
Comment (dzajkowski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: agreed, it's just hard to follow. would it make sense to name the OR'ed steps?

**Review by dzajkowski — APPROVED** [review]

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 7 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:79]: "contain" is correct, because multiple block events are being compared

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 8 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: ok

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 10 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: Can you give an example how this should look like using `cond1`, `cond2` as names for OR's?

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala [hunk 11]
```diff
@@ -146,70 +146,70 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
...
-    "!C 4!"     -> HAD_ITS_MATCH              -> conflictsForNow(S0, F_, C0 | S1),
-    "!C (4!)"   -> VOLATILE_EVENT             -> conflictsForNow(S0, F_ | S0, C0),
-    "!C 4!!"    -> HAD_ITS_MATCH              -> conflictsForNow(S0, F_, C0 | R1),
-    "!C !!X"    -> SAME_POLARITY_MERGE        -> conflictsForNow(S0, R1, C0),
-    "!C !!4"    -> COULD_MATCH_SAME_FALSE     -> conflictsForNow(S0, R1, C0 | F1),
-    "!C CX"     -> HAD_ITS_MATCH              -> conflictsForNow(S0, C_, C_),
-    "!C C!"     -> HAD_ITS_MATCH              -> conflictsForNow(S0, C_, C0 | S1),
+    "!4 4!!"    -> HAD_ITS_MATCH              -> merges(S0, F_, F0 | R1),
+    "!4 !!X"    -> SAME_POLARITY_MERGE        -> merges(S0, R1, F0),
+    "!4 !!4"    -> COULD_MATCH_SAME_CONFLICTS -> conflicts(S0, R1, F_ | F1),
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: All `COULD_MATCH_SAME_CONFLICTS` cases should have just one par so that we're sure there's only the single linear receive that is matched by both deploys. So ` F_` intesad of `F_ | F1`. This is non-deterministic currently.

Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: Notice that's yet another auto-validation rule

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 12]
```diff
@@ -35,54 +37,88 @@ object EstimatorHelper {
...
+    ): Boolean =
+      (for {
+        b1  <- b1Events
+        b2  <- b2Events
+        res = b1.conflicts(b2)
+        // TODO: fail fast
+      } yield (res)).contains(true)
+
+    val b2Ops = tuplespaceEventsPerChannel(b2Events)
+    tuplespaceEventsPerChannel(b1Events)
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: please extract a b1Ops val for easier debugging in the future

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 13]
```diff
@@ -35,54 +37,88 @@ object EstimatorHelper {
...
+      .map {
+        case (k, v) =>
+          (k, channelConflicts(v, b2Ops.get(k).getOrElse(Set.empty)))
+      }
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:100]: The result of this `map` will be useful for debuging as well

File: casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+trait Cardinality
+case object Linear    extends Cardinality
+case object NonLinear extends Cardinality
+
+object TuplespaceEvent {
+
+  implicit private[this] def liftProduce(produce: Produce): TuplespaceOperation =
+    TuplespaceOperation(Send, if (produce.persistent) NonLinear else Linear, produce.hash)
+
+  implicit private[this] def liftConsume(consume: Consume): TuplespaceOperation =
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala]: please use explicit method calls, call them `toOperation`

File: casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    TuplespaceOperation(Receive, if (consume.persistent) NonLinear else Linear, consume.hash)
+
+  def from(produce: Produce): (Blake2b256Hash, TuplespaceEvent) =
+    produce.channelsHash -> TuplespaceEvent(
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala]: join next lines

File: casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+      produce,
+      None
+    )
+
+  def from(consume: Consume): Option[(Blake2b256Hash, TuplespaceEvent)] = consume match {
+    case Consume(singleChannelHash :: Nil, _, _, _) =>
+      Some(
+        singleChannelHash -> TuplespaceEvent(
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala]: ditto

File: casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+      produces: Set[Produce]
+  ): Option[(Blake2b256Hash, TuplespaceEvent)] =
+    comm match {
+      case COMM(consume, produce :: Nil, _) => {
+        val incoming: TuplespaceOperation =
+          if (produces.contains(produce)) produce else consume
+        val matched: Option[TuplespaceOperation] = Some(
+          if (incoming == liftProduce(produce)) consume else produce
+        )
+        Some(
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/TuplespaceEvent.scala]: ditto

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 18]
```diff
@@ -35,54 +37,88 @@ object EstimatorHelper {
...
+                                 dag
+                               )
+        (b1AncestorsMap, b2AncestorsMap) = uncommonAncestorsMap.partition {
+          case (_, bitSet) => bitSet == BitSet(0)
+        }
+        b1Events <- extractBlockEvents[F](b1AncestorsMap.keys.toList)
+        b2Events <- extractBlockEvents[F](b2AncestorsMap.keys.toList)
+        conflictsBecauseOfJoins = extractJoinedChannels(b1Events)
+          .intersect(allChannels(b2Events))
+          .nonEmpty || extractJoinedChannels(b2Events).intersect(allChannels(b1Events)).nonEmpty
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:62]: it'd make sense to reuse the keys of `tuplespaceEventsPerChannel(b1Events)` instead of computing the allChannels(b1Events) separately

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 19]
```diff
@@ -35,54 +37,88 @@ object EstimatorHelper {
...
+      .nonEmpty
+  }
+
+  private[this] def isVolatile(comm: COMM, consumes: Set[Consume], produces: Set[Produce]) =
     !comm.consume.persistent && consumes.contains(comm.consume) && comm.produces.forall(
       produce => !produce.persistent && produces.contains(produce)
     )
 
-  private def buildBlockAncestorChannels[F[_]: Monad: BlockStore](
+  type BlockEvents = (Set[Produce], Set[Consume], Set[COMM])
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: please use a case class

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 20]
```diff
@@ -91,32 +127,46 @@ object EstimatorHelper {
...
+      produceEvents                = allProduceEvents.filterNot(producesInVolatileCommEvents.contains(_))
+      consumeEvents                = allConsumeEvents.filterNot(consumesInVolatileCommEvents.contains(_))
+    } yield (produceEvents, consumeEvents, nonVolatileCommEvents)
 
-      ancestorProduceChannels = produceEvents
-        .filterNot(producesInCommEvents.contains(_))
-        .map(_.channelsHash)
-        .toSet
+  private[this] def extractJoinedChannels(b: BlockEvents): Set[Blake2b256Hash] = {
+    def joinedChannels(consumes: Set[Consume]) =
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:145]: please add whitespace around that method

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 21]
```diff
@@ -91,32 +127,46 @@ object EstimatorHelper {
...
-        .flatMap(_.channelsHashes)
-        .toSet
+  private[this] def tuplespaceEventsPerChannel(
+      b: BlockEvents
+  ): Map[Blake2b256Hash, Set[TuplespaceEvent]] =
+    b match {
+      case (produces, consumes, comms) =>
+        val nonPersistentProducesInComms = comms.flatMap(_.produces).filterNot(_.persistent)
+        val nonPersistentConsumesInComms = comms.map(_.consume).filterNot(_.persistent)
+        val freeProduces                 = produces.filterNot(nonPersistentProducesInComms.contains)
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: produces.diff(nonPersistentProducesInComms)

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala [hunk 22]
```diff
@@ -91,32 +127,46 @@ object EstimatorHelper {
...
+        val commEvents    = comms.flatMap(TuplespaceEvent.from(_, produces))
 
-      ancestorChannels = ancestorProduceChannels ++ ancestorConsumeChannels ++ ancestorCommChannels
-      res              = ancestorChannels.map(_.toByteString)
-    } yield res
+        (produceEvents
+          .combine(consumeEvents)
+          .combine(commEvents))
+          .groupBy(_._1)
+          .mapValues[Set[TuplespaceEvent]](_.map(_._2))
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]:
```
        (produceEvents ++ consumeEvents ++ commEvents)
          .groupBy(_._1)
          .mapValues[Set[TuplespaceEvent]](_.map(_._2))
```

if you insist on using cats, please use `|+|`

**Review by ArturGajowy — CHANGES_REQUESTED** [review]
A couple more comments. I look forward to the deferred amendments in the next PR :)

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 11 above)
Comment (lukasz-golebiewski) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: Yes, I can change that for consistency's sake. This case is only non-deterministic with respect to the order in which `R1` is going to match both `F`'s, but this doesn't make a difference in terms of the merges/conflicts result

Yes, we'll get to auto validation, don't worry ;-)

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 22 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: `combine` comes from before refactoring. `++` will do fine now

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala (hunk 4 above)
Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala]: You're right about the COULD_MATCH_SAME_ thing, and we've handled the nondeterminism issue in the other thread :+1:

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 18 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:62]: This means we'd have to always construct the `tuplespaceEventsPerChannel` map, which is costly. I'd leave this as is, because currently if two blocks conflict on joined channels, we fail fast without constructing it

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 19 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: Yeah, will do

**Review by ArturGajowy — APPROVED** [review]
Good job! :+1:

File: casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala (hunk 10 above)
Comment (ArturGajowy) [reply casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala]: I'm pretty sure this is solved now :)
